iroh/src/magicsock.rs

```diff
@@ -1729,6 +1727,71 @@ enum DiscoBoxError {
     Parse(anyhow::Error),
 }
 
+/// Creates a sender and receiver pair for sending datagrams to the [`RelayActor`].
+///
+/// These includes the waker coordination required to support [`AsyncUdpSocket::try_send`]
+/// and [`quinn::UdpPoller::poll_writable`].
+fn relay_datagram_sender() -> (
+    RelayDatagramSendChannelSender,
+    RelayDatagramSendChannelReceiver,
+) {
+    let (sender, receiver) = mpsc::channel(256);
+    let waker = Arc::new(AtomicWaker::new());
+    let tx = RelayDatagramSendChannelSender {
+        sender,
+        waker: waker.clone(),
+    };
+    let rx = RelayDatagramSendChannelReceiver { receiver, waker };
+    (tx, rx)
+}
+
+/// Sender to send datagrams to the [`RelayActor`].
+///
+/// This includes the waker coordination required to support [`AsyncUdpSocket::try_send`]
+/// and [`quinn::UdpPoller::poll_writable`].
+#[derive(Debug, Clone)]
+struct RelayDatagramSendChannelSender {
+    sender: mpsc::Sender<RelaySendItem>,
+    waker: Arc<AtomicWaker>,
+}
+
+impl RelayDatagramSendChannelSender {
+    fn try_send(
+        &self,
+        item: RelaySendItem,
+    ) -> Result<(), mpsc::error::TrySendError<RelaySendItem>> {
+        self.sender.try_send(item)
+    }
+
+    fn poll_writable(&self, cx: &mut Context) -> Poll<io::Result<()>> {
+        match self.sender.capacity() {
+            0 => {
+                self.waker.register(cx.waker());
+                Poll::Pending
+            }
+            _ => Poll::Ready(Ok(())),
```

Comment:
I really wish this would use the `try_reserve` API instead, and store the `Permit` for sending later..

Reply:
https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.try_reserve

Reply:
Bear in mind that `try_send(&self, item)` and `poll_writeable(&mut self, cx)`.

This becomes pretty unwieldy to make this happen.  The poller could be it's own struct, but it would still have a clone of the sender and waker, plus you'd probably need a channel to send an `mpsc::OwnedPermit` back to the `Sender`.  At this point I'm not sure that version still looks appealing.

Reply:
> try_send(&self, item)

I hate this function signature so much..